### ISSUE TITLE
Apply higher standard fee and random drop policy when blocks are artificially full

### ISSUE THREAD

**Issue by zebambam** [zcash/zcash#4026]
When shielded transactions were first created, the existing fee market of Bitcoin at the time was bypassed, and transactions were also given MAX_PRIORITY so that they would be selected in block templates and mined even though they are larger in size than transparent transactions.

This was done to get shielded transactions working, and was only ever intended as a stop-gap measure and was not intended to have survived as long as it has, and into the realm where Sapling transactions have made shielded transaction generation at much faster speeds a reality. Sapling prescribes no limits on transaction size, instead shifting that concern to the supporting code.

That supporting code should now be changed to prevent issue #3983, first described publicly in the zeroth post in issue #3955.

Nothing in this issue or the work that will ensue stops us from later implementing ticket #3473 or anything like it that better match the product we would like to make when blocks are _naturally_ full. At the moment we are not close to that and so this ticket is to handle the situation only when blocks are _artificially_ full owing to an attacker deliberately trying to deny service to our legitimate users.

When the mempool is filled beyond a certain size, zcashd should:

1. Apply a new standard fee scheme described below
2. Remove any transactions with less than the standard fee
3. If the mempool is still filled beyond a certain size, randomly drop some transactions from the mempool

The new fee scheme should penalize larger transactions on a scale:

`fee = 0.0001 + rate*max(0, size - threshold)^x`

How full should the mempool be before zcashd changes behavior? I propose 2 blocks.

What should the values of rate, threshold and x be? I propose that we balance them so that blocks cost about the same for an attacker to fill up no matter what size transactions they use.

What UX changes will be needed for this scheme to work? I think that users need to know when their transactions won't be mined because of this scheme. How do we do that?

What are the details of the random drop scheme? How often and how many transactions should be dropped? How do we determine what the rates should be?

**Comment by geffenz** [comment]
[UX Review]: There are two main areas of UX effort here:

1. Propagating the changes to wallet developers and implement new UI for those that had fees available in their sending form. This also includes removing any functionality in Zcashd that allowed fee manipulation.
2. Communicating the state of a transaction to a user. This is a big one.
- If a tx has less than the standard fee, they should be alerted immediately - preferably before they believe they sent the tx to the mempool (inline alerts). If possible auto update the fee amount.
- If your tx made it to the mempool but is randomly chosen for deletion, we need to educate the sender somehow. What trackback vectors do we have? We may need to recommend wallet developers and zcashd validate in block +1 their tx made it, and if not prompt a user to retry. This links to rollback UX behavior as well.

(Will values of rate and x be WILDLY fluctuating? Is this why threshold exists or do we need an additional one to prevent changes in values cancelling too many txs?)

**Comment by zebambam** [comment]
Rate and x would be fixed. It being a 'standard' fee means "not a variable fee market". But of course it wouldn't kick in until blocks are already full (for some fixed definition of that), so in essence we're talking about two different standard fee calculations. One being 0.0001 and one being calculated according to the formula above.

**Comment by tromer** [comment]
> * If your tx made it to the mempool but is randomly chosen for deletion, we need to educate the sender somehow. What trackback vectors do we have? We may need to recommend wallet developers and zcashd validate in block +1 their tx made it, and if not prompt a user to retry. This links to rollback UX behavior as well.

Yes, there are many situation where transactions don't end up in the chain, and I'm not sure that a FAQ entry (which is how (see https://github.com/zcash/zcash/issues/1959 was closed) cuts it.

**Comment by daira** [comment]
It is possible for blocks to be full without hitting the `mempooltxcostlimit`, and so we do want to improve the policy applied by `getblocktemplate`. It's not clear that we want to do it by applying a higher standard fee when blocks are full. Other policies such as primarily ordering transactions by age, size, or estimated validation cost, may be more effective in preventing DoS. Random drop is still a good approach, though, and I think we should make the policy for `getblocktemplate` be more consistent with the one applied for mempool limitation. I'll change the title of this ticket accordingly. Let's have a meeting soon to discuss what policy we want.

**Comment by daira** [comment]
See also https://github.com/zcash/zcash/issues/1896 and other issues with the [tx propagation/mining latency](https://github.com/zcash/zcash/issues?q=is%3Aopen+is%3Aissue+label%3A%22tx+propagation%2Fmining+latency%22) label.

**Comment by nuttycom** [comment]
This looks like it may be related to (or could hopefully be closed by) #6344

**Comment by nuttycom** [comment]
Related to #6344
